### PR TITLE
Invalidate Dynamo caches when module hooks change

### LINKED ISSUE #117265: torch.compile silently incorrect with full_backward_pre_hook
```
import torch
import torch.nn as nn

layer = nn.Linear(3, 3, bias=False)
layer.weight = nn.Parameter(torch.eye(3, dtype=torch.float))

def hook(module, grad_output):
    return tuple(go * 2 for go in grad_output)

@torch.compile(backend="aot_eager", fullgraph=True)
def f(x):
    y = layer(x)
    return y

x = torch.randn(3, requires_grad=True)
f(x)
layer.register_full_backward_pre_hook(hook)

x = torch.randn(3, requires_grad=True)
y = f(x)
y.sum().backward()
print(x.grad)
```
Gives different results with or without torch.compile.

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186507

Dynamo skips detailed guards for empty nn.Module hook dictionaries, and global hook dictionaries can be covered only by type guards. Public hook registration and RemovableHandle.remove() mutate those OrderedDicts in place, so the module __dict__ version and existing guards do not change. A compiled graph could therefore be reused after register_full_backward_pre_hook, causing silent incorrect gradients.

Fix this by watching nn.Module hook dictionaries when Dynamo guards them and by notifying the mutation guard whenever public module/global hook registration or removal mutates one of those dictionaries, including secondary with_kwargs and always_call dictionaries. RemovableHandle now accepts a callback that receives the mutated dictionary, preserving the existing weak-reference ownership model.

I considered invalidating only the owning module when hooks are registered, but that misses global hook dictionaries and hook removal through RemovableHandle. Watching the hook dictionaries directly keeps the invalidation tied to the state Dynamo actually guarded.

Fixes #117265

Generated by my agent

Benchmark Results:

Command: python - <<'PY' ... PY microbenchmark for first-call torch.compile time and 20,000 forward hook add/remove operations.

main: compile_median_s=0.022675, hook_add_remove_20000_s=0.036590

patched: compile_median_s=0.021941, hook_add_remove_20000_s=0.088600

Compile time was noise-level similar. Hook add/remove is slower because it now sends Dynamo mutation notifications on public hook mutations.

Test Plan:

python test/dynamo/test_modules.py -k test_register_full_backward_pre_hook_invalidates -k test_register_forward_hook_invalidates -k test_register_global_forward_hook_invalidates -k test_register_forward_hook_secondary_dicts_invalidate -k test_register_global_forward_hook_secondary_dicts_invalidate -k test_register_global_full_backward_pre_hook_invalidates -k test_backward_hooks

python test/nn/test_module_hooks.py -k test_removable_handle_does_not_keep_hook_dict_alive

lintrunner -a

git diff --check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the CI failures. The Dynamo unit failure was a stale assertion that expected guard_fail_fn to run after proactive hook-dict invalidation, so the test now checks the observable output and recompilation behavior instead. The inductor TorchBench graph-break regressions came from running Dynamo hook-dict invalidation notifications while Dynamo was already tracing hook registration/removal; those notifications are now eager-only so in-trace hook mutations stay handled by Dynamo side-effect tracking.

**Comment by jansel** [comment]
@claude review these changes

**Review by zou3519 — COMMENTED** [review]
based on animesh's comment on the issue, it sounds like we just need to flip a few configs on for this to work. Was that not the case?

**Comment by jansel** [comment]
My agent says flipping `inline_inbuilt_nn_modules` / `guard_nn_modules` makes the original repro take the guarded path, but it is a broader mitigation rather than the root fix here. The stale-cache bug is that public hook registration/removal mutates the module/global hook OrderedDicts in place after a graph has already been compiled, so existing guards can keep reusing code that depended on the old hook state. This PR implements the invalidation approach discussed later in the issue: watch the specific hook dicts Dynamo guards and notify Dynamo when hook registration/removal mutates them, including global hooks and secondary hook dicts. That keeps the fix local to hook-state mutations instead of depending on changing those config defaults.

**Review by anijain2305 — CHANGES_REQUESTED** [review]
Changes to nn/modules/module.py seems like a bad idea. We are making nn/modules/module.py codebase aware of the presence of Dynamo. In the past, I have seen resistance in this path.

We should let humans discuss this before proceeding.

**Comment by jansel** [comment]
My agent says I addressed the Dynamo-coupling concern by replacing the direct nn.Module-to-Dynamo path with a compiler-neutral hook-dictionary mutation observer in torch.utils.hooks. nn.Module now only emits generic notifications after public hook mutations; torch._dynamo.mutation_guard subscribes on the compiler side and invalidates the exact guarded dictionary. The rebased change passes the Dynamo hook tests, all 55 module-hook tests, and lintrunner.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
